### PR TITLE
Generate TPU<->CPU switch functions boilerplate

### PR DESCRIPTION
**Pull request by dlibenzi**

### CODE REVIEW

**Comment by dlibenzi** [comment]
Generated code:

https://gist.github.com/dlibenzi/b842b0de1d470037fafd5337e194b43c

File: scripts/gen.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+    fn(t)
+  else:
+    assert isinstance(t, lark.tree.Tree)
+    for c in t.children:
+      for_every_token(c, fn)
+
+
+def emit_string(t, emit, emit_fn):
+  status = emit_fn(t)
+  if status > 0:
```
Comment (asuhan) [inline scripts/gen.py]: When is this branch taken? All `emit_fn` I see return either 0 or -1.

Comment (dlibenzi) [reply scripts/gen.py]: It's a feature that allows to whitelist a whole Tree while enumerating it for emission.

**Comment by dlibenzi** [comment]
I am not sure how to organize the files ATM.
For now we generate a single output file, but we can change it once we have a clearer idea of the full plumbing.
I am also not in love at all with the current map lookup wiring, and the casting from std::function to void* kept in the registration map.

**Comment by dlibenzi** [comment]
New generated code:

https://gist.github.com/dlibenzi/761fcc47cbfec0774c73c4945062d0cc

**Comment by dlibenzi** [comment]
Registration map which BTW is very tricky, as if you miss a single space/punctuation, lookup will fail.
This could be done better IMHO, by generating a cpp/h files with proper std::function<> entries, and instead of map.insert() overrides, having set_XXX() API which can be type checked, and whose call does not involve map lookups.

**Comment by dlibenzi** [comment]
New code:

https://gist.github.com/dlibenzi/58665ebe3704810d98a076ae4a57cad1

File: torch_xla/csrc/tensor_impl.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#include <ATen/Type.h>
+#include <c10/core/TensorImpl.h>
+
+#include "tensor.h"
+
+namespace torch_xla {
+
+// Tensor implementation class used to be fed to the at::Tensor.
+// Its scope is just to handle an XLATensor.
+class TensorImpl : public c10::TensorImpl {
```
Comment (asuhan) [inline torch_xla/csrc/tensor_impl.h]: Rename to `XLATensorImpl`, we shouldn't rely on namespaces to disambiguate.

File: torch_xla/csrc/aten_xla_bridge.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// device.
+std::vector<at::Tensor> CreateXlaTensors(
+    const std::vector<at::Tensor>& tensors, const Device& device);
+
+// Extracts the device out of the XLA tensor. Throws an exception if tensor is
+// not an XLA tensor.
+Device XlaTensorDevice(const at::Tensor& tensor);
+
+static inline Device XlaTensorDevice(const at::TensorList& tensors) {
+  return XlaTensorDevice(tensors[0]);
```
Comment (asuhan) [inline torch_xla/csrc/aten_xla_bridge.h]: I'd do `tensors.at(0)` or assert that `tensors` is not empty.

File: scripts/gen.py [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+  sig, fname, xfname = get_function_signature(tree, orig_sig, gen_fnname)
+  code = '{} {{\n'.format(sig)
+  xla_ref_param = ref_param
+  param_vars = []
+  for p in params:
+    ptype = p.children[0]
+    cptype = type_core(ptype)
+    pname = param_name(p)
+    if cptype == 'TensorList':
+      xname = '_l_{}'.format(pname)
```
Comment (asuhan) [inline scripts/gen.py]: Could we avoid prefixing with underscores? It's not allowed by the C++ standard. Same for double underscores anywhere in the identifier, would be nice if we avoided them.

File: torch_xla/csrc/aten_xla_bridge.h (hunk 3 above)
Comment (dlibenzi) [reply torch_xla/csrc/aten_xla_bridge.h]: SG

File: torch_xla/csrc/tensor_impl.h (hunk 2 above)
Comment (dlibenzi) [reply torch_xla/csrc/tensor_impl.h]: namespaces have been invented to stop people prefixing names, but OK, let's use C in this case ;)

File: scripts/gen.py (hunk 4 above)
Comment (dlibenzi) [reply scripts/gen.py]: SG

**Pull request merged**